### PR TITLE
REST API

### PR DESCRIPTION
**Pull request by Vad1mo**
REST API for Portus to manage Users and Tokens. Furthermore based on this effort here it is possible to add more APIs to Portus.

PR Content and Features:
 - CRUD Users and their application tokens via a REST PAI
 - Swagger Spec/OpenAPI 
 - Documentation

### CODE REVIEW

**Comment by mssola** [comment]
First of all, thanks a lot for this 👍 

Now, you may want to take a look at what we are doing here: https://github.com/SUSE/Portus/pull/1289. There's a lot of stuff on this PR, but take into consideration:

- The usage of `PORTUS-AUTH` in  `app/controllers/application_controller.rb`. This is inspired by what Gitlab is doing, and it allows remote clients to make API calls.
- The `app/controllers/tags_controller.rb` file now implements the `show` method, which is the first one providing API support.

Now, this is far from perfect, so I encourage you to innovate upon these ideas. I also thought of using something like [grape](https://github.com/ruby-grape/grape), but so far I've used good ol' Rails as I'm already used to it (also notice all the ugly `as_json` methods, that I'd like to get rid off with a proper serializer or other solution...).

**Comment by Vad1mo** [comment]
I saw some of the work in the branch. We will take PORTUS-AUTH into consideration.  I like the approach that grape takes. We also see this as an viable option.

**Comment by Vad1mo** [comment]
It would be nice if someone could review the API Docs f1999b7 It would be nice if @mssola or someone else would take some time to review the API concept.

**Comment by mssola** [comment]
> I saw some of the work in the branch. We will take PORTUS-AUTH into consideration. @mssola when do you plan to merge #1289 into master?

Both me and @vitoravelino have been busy with other stuff, and it will probably take a while, since we have to review the code, and apply more testing... I'd say I can break away the code for managing `PORTUS-AUTH` in another PR, so you can use it.

> I like the grape approach. We also see this as an viable option and thinking about using it.

I have to admit that I've never used it, so this is something I'm not really sure. It needs some exploration from my part 😄 

> It would be nice if someone could review the API Docs f1999b7 It would be nice if @mssola or someone else would take some time to review the API concept.

Left some comments. In principle it looks good to me 👍

**Comment by mssola** [comment]
@Vad1mo ping me whenever this is ready to be reviewed.

**Comment by Vad1mo** [comment]
We are fixing the travis build and adding some documentation next to the swagger spec. The code is basically ready for an review. 

fyi: you can get the swagger spec localhost:3000/api/swagger_doc

**Comment by mssola** [comment]
@Vad1mo ok, I'll wait also for the swagger documentation if you don't mind 😉

**Comment by andrew2net** [comment]
@mssola Hello,
I'm trying to pass travis. Running command, which got an error on travis:
```
bundle exec rspec spec packaging/suse/portusctl/spec
```
got
```
...
rpc error: code = 2 desc = oci runtime error: exec failed: container_linux.go:247: starting container process caused "exec: \"mysqladmin\": executable file not found in $PATH"

Command 'mysqladmin --count=1 --connect-timeout=1 --host=integration_db -u root --password=portus ping' failed! Waiting 5 seconds...
...
```
can you help me find cause?

**Comment by mssola** [comment]
@andrew2net have you tried without running the integration tests ? (they might be broken...):

    $ bundle exec rspec spec packaging/suse/portusctl/spec --tag ~integration

**Comment by andrew2net** [comment]
@mssola I haven't. traied now and all tests passed. Does it need to pass integration tests for successfull travis checkung?

**Comment by mssola** [comment]
@andrew2net no. Travis does not run integration tests for Portus.

**Comment by Vad1mo** [comment]
@mssola you can take a look now at the PR

File: .gitignore [hunk 1]
```diff
@@ -11,4 +11,5 @@
 /docker-compose.overlay.yml
 /spec/integration/fixtures/config.yml
 /spec/integration/fixtures/data
-/node_modules
\ No newline at end of file
+/node_modules
+.idea
```
Comment (mssola) [inline .gitignore]: Remove this line

File: .rubocop.yml [hunk 2]
```diff
@@ -29,6 +29,9 @@ Style/Documentation:
 Lint/UnneededDisable:
   Enabled: false
 
+Style/FrozenStringLiteralComment:
+  Enabled: false
```
Comment (mssola) [inline .rubocop.yml]: This wasn't an issue before 😕 If you call it with `bundle exec` there shouldn't be any problem no ?

File: .ruby-version [hunk 3]
```diff
@@ -1 +1 @@
-2.1.2
+2.3.3
```
Comment (mssola) [inline .ruby-version]: Unrelated to this PR.

File: .travis.yml [hunk 4]
```diff
@@ -1,11 +1,7 @@
 language: ruby
 rvm:
-  # The version used at SLE12
-  - 2.1.2
 
   # Stable versions
-  - 2.1.10
-  - 2.2.6
```
Comment (mssola) [inline .travis.yml]: The changes on this file are unrelated to this PR. Moreover, why should we stop testing these versions ? (maybe we could upgrade the 2.2.6 to 2.2.7)

File: Gemfile [hunk 5]
```diff
@@ -22,6 +22,12 @@ gem "redcarpet"
 gem "font-awesome-rails"
 gem "rails_stdout_logging", "~> 0.0.5", group: [:development, :staging, :production]
 gem "webpack-rails"
+gem "grape"
+gem "grape-entity"
+gem "grape-swagger"
+gem "grape-swagger-entity"
+gem "grape-swagger-representable"
+gem "hashie-forbidden_attributes"
```
Comment (mssola) [inline Gemfile]: I'm not sure about this gem... Do you *really* need mass assignment ?

File: Gemfile.lock [hunk 6]
```diff
@@ -448,4 +506,4 @@ DEPENDENCIES
   wirble
 
 BUNDLED WITH
-   1.13.6
+   1.15.1
```
Comment (mssola) [inline Gemfile.lock]: I'm not sure about all the changes on the `Gemfile.lock`. Are you sure you only run `bundle install grape`, and not `bundle update` ? I'd like to keep the changes on versions as minimal as possible, since we will have to maintain them 😉

File: REST-API.md [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+
+## OpenAPI Specification
+
+OpenAPI Specification (fka The Swagger Specification) can be fetched from the running Portus instance by visiting the URL `/api/swagger_doc` e.g. 
+[localhost:3000/api/swagger_doc](http://localhost:3000/api/swagger_doc)
+
+To display the results in a more human readable form you can use the [Swagger UI](http://petstore.swagger.io/) and import the above URL. 
+ 
+
```
Comment (mssola) [inline REST-API.md]: I'd move this document into the `docs` directory and simply name it `API.md`.

File: REST-API.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+```
+request.headers["PORTUS-AUTH"]: "username:token"
+```
+If the authentication fails, then the HTTP status code 401 is returned. If authorization fails a 403 status code is returned.
+
+
+## OpenAPI Specification
+
+OpenAPI Specification (fka The Swagger Specification) can be fetched from the running Portus instance by visiting the URL `/api/swagger_doc` e.g.
```
Comment (mssola) [inline REST-API.md]: I'd start we "The". Moreover there's also a typo: fka should be aka.

File: REST-API.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+If the authentication fails, then the HTTP status code 401 is returned. If authorization fails a 403 status code is returned.
+
+
+## OpenAPI Specification
+
+OpenAPI Specification (fka The Swagger Specification) can be fetched from the running Portus instance by visiting the URL `/api/swagger_doc` e.g. 
+[localhost:3000/api/swagger_doc](http://localhost:3000/api/swagger_doc)
+
+To display the results in a more human readable form you can use the [Swagger UI](http://petstore.swagger.io/) and import the above URL. 
+
```
Comment (mssola) [inline REST-API.md]: Extra empty line (nitpicking 😁)

File: spec/models/namespace_spec.rb [hunk 10]
```diff
@@ -7,10 +7,9 @@
...
 #  team_id     :integer
-#  registry_id :integer          not null
+#  public      :boolean          default("0")
+#  registry_id :integer
 #  global      :boolean          default("0")
-#  description :text(65535)
-#  visibility  :integer
 #
 # Indexes
 #
```
Comment (mssola) [inline spec/models/namespace_spec.rb]: Changes unrelated to this PR

File: spec/factories/namespaces.rb [hunk 11]
```diff
@@ -7,10 +7,9 @@
...
 #  updated_at  :datetime         not null
 #  team_id     :integer
-#  registry_id :integer          not null
+#  public      :boolean          default("0")
+#  registry_id :integer
 #  global      :boolean          default("0")
-#  description :text(65535)
-#  visibility  :integer
 #
 # Indexes
```
Comment (mssola) [inline spec/factories/namespaces.rb]: Changes unrelated to this PR

File: config/routes.rb [hunk 12]
```diff
@@ -23,6 +23,7 @@
   resources :repositories, only: [:index, :show, :destroy] do
     post :toggle_star, on: :member
     resources :comments, only: [:create, :destroy]
+    resources :vulnerabilities, only: [:index]
```
Comment (mssola) [inline config/routes.rb]: This is from another PR. Let's avoid mixing it here 😉

File: config/routes.rb [hunk 13]
```diff
@@ -52,6 +53,8 @@
     end
   end
 
+  mount GrapeAPI::RootAPI => "/"
```
Comment (mssola) [inline config/routes.rb]: I see that with this we have the following:

- Files have to go in `app/api/grape_api`, while I'd put it in `lib/api`.
- It's not versioned.

Take a look at Gitlab's `lib/api/api.rb` file, since it's closer to what I'd want to accomplish.

File: config/application.rb [hunk 14]
```diff
@@ -39,5 +39,9 @@ class Application < Rails::Application
     config.webpack.output_dir  = "public/assets/webpack"
     config.webpack.public_path = "assets/webpack"
     config.webpack.dev_server.enabled = false
+
+    # Loading Grape API framework.
+    config.paths.add File.join("app", "api"), glob: File.join("**", "*.rb")
+    config.autoload_paths += Dir[Rails.root.join("app", "api", "*")]
```
Comment (mssola) [inline config/application.rb]: Not really needed if you put it inside of `lib/api` for example. Otherwise, you should add it in the same way it's done above for the `lib` and `app/validators` directories...

File: config/initializers/reload_api.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+if Rails.env.development?
```
Comment (mssola) [inline config/initializers/reload_api.rb]: Why do we need this ?

File: app/models/registry_event.rb [hunk 16]
```diff
@@ -3,7 +3,7 @@
 # Table name: registry_events
 #
 #  id         :integer          not null, primary key
-#  event_id   :string(255)      default(""), not null
+#  event_id   :string(255)      default("")
```
Comment (mssola) [inline app/models/registry_event.rb]: Unrelated to PR.

File: app/models/namespace.rb [hunk 17]
```diff
@@ -7,10 +7,9 @@
 #  created_at  :datetime         not null
 #  updated_at  :datetime         not null
 #  team_id     :integer
-#  registry_id :integer          not null
+#  public      :boolean          default("0")
+#  registry_id :integer
 #  global      :boolean          default("0")
-#  description :text(65535)
-#  visibility  :integer
```
Comment (mssola) [inline app/models/namespace.rb]: Unrelated to PR.

File: app/controllers/vulnerabilities_controller.rb [hunk 18]
```diff
@@ -0,0 +1,20 @@
+class VulnerabilitiesController < ApplicationController
```
Comment (mssola) [inline app/controllers/vulnerabilities_controller.rb]: This is from another PR. Please, don't include it...

File: app/api/grape_api/user_api.rb [hunk 19]
```diff
@@ -0,0 +1,191 @@
+module GrapeAPI
```
Comment (mssola) [inline app/api/grape_api/user_api.rb]: Two things:

- There's functionality already existing in `app/controllers/application_tokens_controller.rb` for this. Could we merge the code somehow ? (either changes the frontend to use this API instead of the controller, thus removing the controller, or creating methods that both the controller and the API use)
- Could you add more documentation ?

File: app/api/grape_api/root_api.rb [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+      }
+    end
+
+    class Messages < Grape::Entity
+      expose :message
+    end
+  end
+
+  class RootAPI < Grape::API
+    # version 'v1', using: :path
```
Comment (mssola) [inline app/api/grape_api/root_api.rb]: why not? 😀

File: app/api/grape_api/root_api.rb [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+    helpers do
+      def authenticate!
+        return if request.request_method == "OPTIONS"
+        error!("Authentication fails.", 401) unless (user = current_user)
+        error!("Authorization fails.", 403) unless user.admin
+      end
+
+      def current_user
```
Comment (mssola) [inline app/api/grape_api/root_api.rb]: This already exists in the `application_controller.rb`. We should be using the same code and share it somehow.

File: app/api/grape_api/root_api.rb [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+          name: "Portus-Auth",
+          in:   "header"
+        }
+      },
+      security:             [api_key: []],
+      info:                 {
+        title:         "Portus user API",
+        description:   "CRUD users and user's application tokens API",
+        contact_name:  "Anldrei Kislichenko",
+        contact_email: "andrei.kislichenko@gmail.com"
```
Comment (mssola) [inline app/api/grape_api/root_api.rb]: I'd prefer:

- Contact name: "Portus authors" (which would include you 😉)
- Contact email: "portus-dev@googlegroups.com"

File: .rubocop.yml (hunk 2 above)
Comment (andrew2net) [inline .rubocop.yml]: @mssola with ruby 2.3 rubocop reports offense `Missing frozen string literal comment.` on each rb file

File: Gemfile (hunk 5 above)
Comment (andrew2net) [inline Gemfile]: @mssola ruby-grape uses it https://github.com/ruby-grape/grape#rails

File: Gemfile.lock (hunk 6 above)
Comment (andrew2net) [inline Gemfile.lock]: @mssola I didn't run `bundle update`, just updated bundle because of warning, it possible to return v 1.13.6

File: config/initializers/reload_api.rb (hunk 15 above)
Comment (andrew2net) [inline config/initializers/reload_api.rb]: @mssola this reload API changes https://github.com/ruby-grape/grape#reloading-api-changes-in-development

File: config/routes.rb (hunk 12 above)
Comment (andrew2net) [inline config/routes.rb]: @mssola we did cherry-pick "Added an authentication layer for the API" commit bcd9c65 from #1299. so some changes go with it. I'll remove them.

File: app/api/grape_api/root_api.rb (hunk 22 above)
Comment (andrew2net) [inline app/api/grape_api/root_api.rb]: @mssola ok, I don't mind 😄

File: .ruby-version (hunk 3 above)
Comment (Vad1mo) [inline .ruby-version]: well, grape requires to go to minimum 2.3.

File: .travis.yml (hunk 4 above)
Comment (Vad1mo) [inline .travis.yml]: ruby 2.3 is required

File: spec/models/namespace_spec.rb (hunk 10 above)
Comment (Vad1mo) [inline spec/models/namespace_spec.rb]: I don't know how others a handling this but the comments are automatically updated by the annotate-models gem.  I can make a new PR and update all file so they are in sync with the gem default.

File: config/routes.rb (hunk 13 above)
Comment (Vad1mo) [inline config/routes.rb]: app/api/ is the grape default. If you want to setup differently we can try move it to lib and see if that works.

File: app/api/grape_api/user_api.rb (hunk 19 above)
Comment (andrew2net) [inline app/api/grape_api/user_api.rb]: @mssola the functionality slightly different. API can create tokens for any user, not only for current, and return plain_token. I think changing the frontend to use the API is preferred.
Will add more doc.

File: app/api/grape_api/root_api.rb (hunk 20 above)
Comment (andrew2net) [inline app/api/grape_api/root_api.rb]: indeed 😄

File: app/api/grape_api/root_api.rb (hunk 21 above)
Comment (andrew2net) [inline app/api/grape_api/root_api.rb]: Current_user is in Devise gem. It is not the same function. I'll rename it to avoid confusion.

File: app/api/grape_api/user_api.rb (hunk 19 above)
Comment (andrew2net) [inline app/api/grape_api/user_api.rb]: The functionality is moved to model.

File: .rubocop.yml (hunk 2 above)
Comment (mssola) [inline .rubocop.yml]: Gotcha

File: .ruby-version (hunk 3 above)
Comment (mssola) [inline .ruby-version]: I'll have to check about this requirement...

File: spec/models/namespace_spec.rb (hunk 10 above)
Comment (mssola) [inline spec/models/namespace_spec.rb]: Unfortunately yes. It's a side-effect of one of the gems... Maybe you can simply `checkout --` these changes before merging 😉

**Comment by mssola** [comment]
Note that we have just merged #1289, and so there are quite some conflicts now. Also, make sure to rebase against master.

**Comment by andrew2net** [comment]
@mssola Hi,
We rebased to master. Seems it has Javascript changes. Then I try to run ```rake webpack:compile``` I got
```
...
/home/vagrant/RubyProjects/Portus/node_modules/.bin/webpack --config /home/vagrant/RubyProjects/Portus/config/webpack.js --bail
loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
ModuleNotFoundError: Module not found: Error: Can't resolve 'vue-resource' in '/home/vagrant/RubyProjects/Portus/app/assets/javascripts'
    at factoryCallback (/home/vagrant/RubyProjects/Portus/node_modules/webpack/lib/Compilation.js:259:39)
...
```
can you help to resolve the issue?

**Comment by andrew2net** [comment]
@mssola Hello Miquel,
we rebased an api branch against a master, but I can't figure out why some files from the master aren't covered by tests in api. when rspec runs with master test coverage is 100%, with api less than 100%.
tell me, please where are specs for following files?
```
app/controllers/vulnerabilities_controller.rb
app/controllers/webhook_deliveries_controller.rb
app/policies/webhook_delivery_policy.rb
app/serializers/namespace_serializer.rb
app/serializers/repository_serializer.rb
lib/man_pages.rb
```
or, maybe the files excluded somehow from tests coverage calculation?

coverage

**Comment by Vad1mo** [comment]
@mssola IMHO, we are ready now.

File: lib/api/root_api.rb [hunk 23]
```diff
@@ -0,0 +1,96 @@
+require "grape-swagger"
+
+module API
+  module Entities
+    class Users < Grape::Entity
+      expose :id, documentation: { type: Integer, desc: "User id" }
+      expose :username, documentation: { type: String, desc: "User name" }
+      expose :email, documentation: { type: String, desc: "E-mail" }
+      expose :sign_in_count, documentation: { type: Integer }
```
Comment (mssola) [inline lib/api/root_api.rb]: Is this really needed ?

File: lib/api/root_api.rb [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+      expose :current_sign_in_at, documentation: { type: DateTime }
+      expose :last_sign_in_at, documentation: { type: DateTime }
+      expose :current_sign_in_ip, :last_sign_in_ip
```
Comment (mssola) [inline lib/api/root_api.rb]: Not sure about these two, since it's not that useful imho

File: lib/api/root_api.rb [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+      expose :created_at, :updated_at, documentation: { type: DateTime }
+      expose :admin, :enabled, documentation: { type: "boolean" }
+      expose :ldap_name, documentation: { type: String, desc: "Ldap name" }
```
Comment (mssola) [inline lib/api/root_api.rb]: This is a legacy column. You can safely skip it.

File: lib/api/root_api.rb [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+      expose :failed_attempts, documentation: { type: Integer }
```
Comment (mssola) [inline lib/api/root_api.rb]: I wouldn't expose this one either.

File: lib/api/root_api.rb [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+        type: "API::Entities::Messages", is_array: true
+      }
+    end
+
+    class Messages < Grape::Entity
+      expose :message
+    end
+  end
+
+  # Load user api for ralke task.
```
Comment (mssola) [inline lib/api/root_api.rb]: Typo: "ralke" -> "rake"

File: lib/api/root_api.rb [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+        }
+      },
+      security:             [api_key: []],
+      info:                 {
+        title:         "Portus user API",
+        description:   "CRUD users and user's application tokens API",
+        contact_name:  "Portus authors",
+        contact_email: "portus-dev@googlegroups.com"
+      },
+      schemes:              [:http, :https]
```
Comment (mssola) [inline lib/api/root_api.rb]: You can further restrict this, no ? For example, you can check if `PORTUS_CHECK_SSL_USAGE_ENABLED` is set to true (which is set in the production Docker image). If so, then only set `[:https]` as the allowed schemes.

File: lib/api/v1/user_api.rb [hunk 29]
```diff
@@ -0,0 +1,200 @@
+module API
+  module V1
+    class UserAPI < Grape::API
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: I think it's more aligned with rails if:

- The file is named `users.rb`.
- The class is named Users.

Plural because it's a resource, and without the `API` part because that's already stated by the name of the first module.

File: lib/api/v1/user_api.rb [hunk 30]
```diff
@@ -0,0 +1,200 @@
...
+      version "v1", using: :path
+
+      resource :users do
+        before do
+          authorization!
+        end
+
+        route_param :id, type: Integer do
+          resource :application_tokens do
+
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: Remove empty line

File: lib/api/v1/user_api.rb [hunk 31]
```diff
@@ -0,0 +1,200 @@
...
+              failure:  [
+                [400, "Bad request.", API::Entities::ApiErrors],
+                [401, "Authentication fails."],
+                [403, "Authorization fails."]
+              ],
+              consumes: ["application/x-www-form-urlencoded"]
+
+            params do
+              requires :application, documentation: { desc: "Application name" }
+            end
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: Add an empty line after the `end` line.

File: lib/api/v1/user_api.rb [hunk 32]
```diff
@@ -0,0 +1,200 @@
...
+          desc "Delete application token.",
+            failure: [
+              [401, "Authentication fails."],
+              [403, "Authorization fails."],
+              [404, "Not found."]
+            ]
+
+          params do
+            requires :id, documentation: { desc: "Token id" }
+          end
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: Same here

File: lib/api/v1/user_api.rb [hunk 33]
```diff
@@ -0,0 +1,200 @@
...
+          requires :user, type: Hash do
+            requires :all,
+              only:  [:username, :email],
+              using: API::Entities::Users.documentation.slice(:username, :email)
+            requires :password, type: String
+            optional :all,
+              only:  [:display_name],
+              using: API::Entities::Users.documentation.slice(:display_name)
+          end
+        end
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: Same here

File: lib/api/v1/user_api.rb [hunk 34]
```diff
@@ -0,0 +1,200 @@
...
+          requires :user, type: Hash do
+            optional :all,
+              only:  [:username, :email],
+              using: API::Entities::Users.documentation.slice(:username, :email)
+            optional :password, type: String, desc: "Password"
+            optional :all,
+              only:  [:display_name],
+              using: API::Entities::Users.documentation.slice(:display_name)
+          end
+        end
```
Comment (mssola) [inline lib/api/v1/user_api.rb]: Same

File: lib/api/root_api.rb (hunk 23 above)
Comment (Vad1mo) [inline lib/api/root_api.rb]: `sign_in_count` we can skip. The others I would like to keep.

File: lib/api/root_api.rb (hunk 24 above)
Comment (Vad1mo) [inline lib/api/root_api.rb]: ok, not really relevant.

File: lib/api/root_api.rb (hunk 25 above)
Comment (Vad1mo) [inline lib/api/root_api.rb]: ok,

File: lib/api/root_api.rb (hunk 26 above)
Comment (Vad1mo) [inline lib/api/root_api.rb]: ok,

File: lib/api/v1/users.rb [hunk 35]
```diff
@@ -0,0 +1,204 @@
...
+          end
+        end
+
+        desc "Create new user.",
+          failure:  [
+            [400, "Bad request.", API::Entities::ApiErrors],
+            [401, "Authentication fails."],
+            [403, "Authorization fails."]
+          ],
+          consumes: ["application/x-www-form-urlencoded"]
```
Comment (mssola) [inline lib/api/v1/users.rb]: Maybe this can be done in another PR, but would it be too difficult to also support JSON ? It would be simpler than with an encoded form. See the curl command I had to use:

```
curl -X POST \
     -H "PORTUS-AUTH: mssola:KXsgWZQyDq8rMoy-hHaD" \
     -F "user[username]=another" -F "user[email]=asd@example.org" -F "user[password]=12341234" \
     https://registry.mssola.cat/api/v1/users
```

Maybe this could work too (so without the `user[...]` part):

```
curl -X POST \
     -H "PORTUS-AUTH: mssola:KXsgWZQyDq8rMoy-hHaD" \
     https://registry.mssola.cat/api/v1/users?username=another&email=asd@example.org&password=12341234
```

File: lib/api/v1/users.rb [hunk 36]
```diff
@@ -0,0 +1,204 @@
...
+        # Update user with given :id.
+        desc "Update user.",
+          params:   API::Entities::Users.documentation.slice(:id),
+          failure:  [
+            [400, "Bad request.", API::Entities::ApiErrors],
+            [401, "Authentication fails."],
+            [403, "Authorization fails."],
+            [404, "Not found."]
+          ],
+          consumes: ["application/x-www-form-urlencoded"]
```
Comment (mssola) [inline lib/api/v1/users.rb]: Same here.

**Comment by andrew2net** [comment]
@mssola API with GRAPE processes requests if the path matches any of endpoints. If not then the request is ignored by GRAPE and the app processes it and returns ``No route matches`` error in development as HTML if route not exists.

File: lib/api/root_api.rb [hunk 37]
```diff
@@ -0,0 +1,93 @@
...
+      security_definitions: {
+        api_key: {
+          type: "apiKey",
+          name: "Portus-Auth",
+          in:   "header"
+        }
+      },
+      security:             [api_key: []],
+      info:                 {
+        title:         "Portus user API",
```
Comment (andrew2net) [inline lib/api/root_api.rb]: Ok

File: lib/api/root_api.rb [hunk 38]
```diff
@@ -0,0 +1,93 @@
...
+        description:   "CRUD users and user's application tokens API",
```
Comment (andrew2net) [inline lib/api/root_api.rb]: Ok

File: lib/api/v1/users.rb (hunk 35 above)
Comment (andrew2net) [inline lib/api/v1/users.rb]: Will add JSON

File: lib/api/v1/users.rb (hunk 36 above)
Comment (andrew2net) [inline lib/api/v1/users.rb]: Ok

File: spec/api/grape_api/v1/users_spec.rb [hunk 39]
```diff
@@ -0,0 +1,233 @@
...
+
+    context "with invalid params" do
+      it "returns errors" do
+        post "/api/v1/users", { user: { username: "", email: "", password: "" } },
+          @header
+        expect(response).to have_http_status(:bad_request)
+        expect(JSON.parse(response.body)["errors"]).not_to be_nil
+      end
+
+      it "returs user error" do
```
Comment (mssola) [inline spec/api/grape_api/v1/users_spec.rb]: Typo: "returs" -> "returns"
